### ISSUE TITLE
tf.pow(x, y) will freeze for negative integer y

### ISSUE THREAD

**Issue by korepwx** [tensorflow/tensorflow#9560]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10, and macOS 10.12.4
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.1.0 on both operation systems
- **Bazel version (if compiling from source)**: not compiled from source
- **CUDA/cuDNN version**: CPU only on Windows, while CUDA 8.0 on macOS
- **GPU model and memory**: Nvidia Titan X, 12GB
- **Exact command to reproduce**: a short piece of Python code
- **Python version**: 3.5.2 (Anaconda) on Windows, 3.6.0 (Anaconda) on macOS

### Describe the problem

tf.pow(x, y) will freeze for negative integer y (and of course, integer x).  It will not freeze for negative floating number y.

### Source code / logs

```python
import tensorflow as tf

with tf.Graph().as_default(), tf.Session().as_default():
    print(tf.pow(5, -2).eval())  # this will not stop
```

**Comment by aselle** [comment]
I've confirmed and reproduced this, thanks for reporting.

**Comment by yongtang** [comment]
I think this issue is related to issue #12156. Also related to PR #11852.

**Comment by codrut3** [comment]
I've looked in the source code and the origin of this bug is in Eigen, specifically src/Core/MathFunctions.h. They specialize their template for pow to the case of both parameters integers, and provide an implementation which only works for positive y. The problem is the well-known bug-loop of shifting y to the left: if y < 0, then the bit sign extends and it continues indefinitely.

Here is the code from Eigen:
`template<typename ScalarX,typename ScalarY>
struct pow_impl<ScalarX,ScalarY, true>
{
  typedef ScalarX result_type;
  static EIGEN_DEVICE_FUNC inline ScalarX run(ScalarX x, ScalarY y)
  {
    ScalarX res(1);
    eigen_assert(!NumTraits<ScalarY>::IsSigned || y >= 0);
    if(y & 1) res *= x;
    y >>= 1;
    while(y)
    {
      x *= x;
      if(y&1) res *= x;
      y >>= 1;
    }
    return res;
  }
};`

However, it would be difficult to fix this: the template specifically requests that the return type is Scalar, and not Real, and changing this in Eigen might break things for other people.

What I could do, would be to change pow in cwise_ops.h in tensorflow/core to fix this bug. But then again, tf.pow would start returning float for integers inputs. Is this acceptable? I don't see any way to fix the bug without accepting that tf.pow must also return float for integer inputs.

If you agree with this change, I will fix the issue.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yongtang** [comment]
PR #15607 is pending review. There are discussions about fixing it in Eigen vs. TensorFlow. Will take a look at Eigen.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yongtang** [comment]
@korepwx I think the issue is addressed in PR #15607 and this issues could be closed.
